zebra-chain/src/work/tests/prop.rs

```diff
@@ -129,3 +130,205 @@ fn equihash_prop_test_input() -> color_eyre::eyre::Result<()> {
 
     Ok(())
 }
+
+proptest! {
+    /// Check Expanded, Compact, Work, and PartialCumulativeWork conversions.
+    ///
+    /// Make sure the conversions don't panic, and that they round-trip and compare
+    /// correctly.
+    #[test]
+    fn prop_difficulty_conversion(expanded_seed in any::<block::Hash>()) {
+        let expanded_seed = ExpandedDifficulty::from_hash(&expanded_seed);
+
+        let hash_zero = block::Hash([0; 32]);
+        let hash_max = block::Hash([0xff; 32]);
+        prop_assert!(expanded_seed >= hash_zero);
+        prop_assert!(expanded_seed <= hash_max);
+
+        // Skip invalid seeds
+        prop_assume!(expanded_seed != hash_zero);
+
+        let compact = expanded_seed.to_compact();
+        let expanded_trunc = compact.to_expanded();
+        let work = compact.to_work();
+
+        let work_zero = Work(0);
+        let work_max = Work(u128::MAX);
+
+        if let Some(expanded_trunc) = expanded_trunc {
```

Comment:
I'm worried that we're not ever hitting this branch or other similar branches like the one below for work, and that some of the proptests based on `filter`s aren't actually going to trigger before they exceed the reject count. We should verify that all of these proptests are working as expected.

Reply:
How do you know that there will be a large number of `work` rejections?

Here's how `work` is calculated:
```
work = 2^256 / expanded
```

So if the original `expanded` value is large (and most uniformly distributed values are large), the corresponding `work` value should be small.

Specifically, I expect an distribution of `work` values similar to:
```
1: 1/2 of all values
2: 1/2-1/3 = 1/6 of all values
3: 1/3-1/4 = 1/12 of all values
4: 1/4-1/5 = 1/20 of all values
...
```

But it's easy enough to move the `Work` checks into functions that generate `any::<Work>`. (Or move the non-`Work` checks before any `prop_assume!`s.) So I can just do that as a precaution.

Reply:
Split Work from other proptests in 808f6317e62b8025b3c3ae427d6df2e26ccc51cd.

I don't know how to verify if proptests are working.

I tried setting `PROPTEST_MAX_LOCAL_REJECTS` and `PROPTEST_MAX_GLOBAL_REJECTS` to 0 and 1, and the tests still passed.

But I don't know if "abort" means "fail the test" or "skip the test".

https://docs.rs/proptest/0.10.1/proptest/test_runner/struct.Config.html#structfield.max_local_rejects

Reply:
sounds like its good, I'd have just put some tracing events into each of the branches to ensure that they were emitting output, but I think if it passes with multiple iterations and minimal rejects allowed, which afaik counter filtered out inputs, then I think we're good.